Add java compiler compatibility fix agains module opens

hegyibalint opened the pull request:
Adds a forward and backward compatibility fix caused by some new modules required:
 - The backward direction is done by the compiler, which will switch to forking mode if the required modules are not open.
 - The forward direction is done by adding the missing module declarations in TestKit

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

big-guy reviewed (changes requested):
Based on what we talked about, we're going to try a different direction

hegyibalint commented on platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AbstractJavaCompileSpecFactory.java:
```diff
@@ -16,53 +16,49 @@
 
     @Override
     public T create() {
-        if (toolchain != null) {
```
ℹ️ Turns out `toolchain` was never null.

hegyibalint commented on platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AbstractJavaCompileSpecFactory.java:
```diff
@@ -78,11 +74,7 @@ private T chooseSpecFromCompileOptions(File fallbackJavaHome) {
         }
 
         final int languageVersion;
-        if (toolchain != null) {
-            languageVersion = toolchain.getLanguageVersion().asInt();
-        } else {
-            languageVersion = JavaVersion.current().ordinal() + 1;
-        }
```
ℹ️ Turns out toolchain was never null, only retained the non-null path.

big-guy commented on platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AbstractJavaCompileSpecFactory.java:
```diff
@@ -16,53 +16,51 @@
     }
 
     @Override
     public T create() {
```
I think we should twist the method around to make forking the default path and in-process execution an exception, I think something like this:
```
    @Override
    public T create() {
        File toolchainJavaHome = toolchain.getInstallationPath().getAsFile();
        if (!toolchain.getLanguageVersion().canCompileOrRun(8)) {
            LOGGER.info("Falling back to command line compiler: requested toolchain is below JDK 1.7");
            return getCommandLineSpec(Jvm.forHome(toolchainJavaHome).getJavacExecutable());
        }

        if (compileOptions.isFork()) {
            File forkJavaHome = compileOptions.getForkOptions().getJavaHome();
            if (forkJavaHome != null) {
                LOGGER.info("Falling back to command line compiler: forking with Java Home set");
                return getCommandLineSpec(Jvm.forHome(forkJavaHome).getJavacExecutable());
            }

            String forkExecutable = compileOptions.getForkOptions().getExecutable();
            if (forkExecutable != null) {
                LOGGER.info("Falling back to command line compiler: forking with javac executable set");
                return getCommandLineSpec(JavaExecutableUtils.resolveExecutable(forkExecutable));
            }
        }

        if (toolchain.isCurrentJvm() && JdkJavaCompiler.canBeUsed()) {
            // Please keep it in mind, that when using TestKit with debug enabled (i.e. in embedded mode), this line won't be reached after Java 16 (JEP 396)
            // If you need this to be executed, add the necessary configs from JPMSConfiguration to the test runner executing Gradle
            LOGGER.info("Using in-process compilation");
            return getDefaultSpec();
        }

        LOGGER.info("Running compilation in compiler daemon");
        return getForkingSpec(toolchainJavaHome, toolchain.getLanguageVersion().asInt());
    }
```

big-guy replied:
Separately, we could rename `getDefaultSpec` to `getInProcessSpec` or similar.

hegyibalint replied:
I agree that it seems like `getForkingSpec` became a go-to fallback, and it should work in all circumstances.

hegyibalint replied:
> I think we should twist the method around to make forking the default path and in-process execution an exception, I think something like this...

Agreed, and used your version.

> Separately, we could rename getDefaultSpec to getInProcessSpec or similar.

Done

big-guy reviewed (changes requested):
A couple of minor comments. 

Are you going to squash these commits?  We probably don't care about the reverts/earlier path you took.

hegyibalint commented:
> Are you going to squash these commits? We probably don't care about the reverts/earlier path you took.

I've rebased the branch, removed the reverts, but tried to keep some logical separation between the commits.

big-guy commented on platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AbstractJavaCompileSpecFactory.java:
```diff
@@ -16,54 +16,58 @@
 
-        if (!toolchain.isCurrentJvm()) {
-            return getForkingSpec(toolchainJavaHome, toolchain.getLanguageVersion().asInt());
+        if (toolchain.isCurrentJvm() && JdkJavaCompiler.canBeUsed()) {
```
Ah, I got this slightly wrong.  This now allows things to run in process when compileOptions.fork = true and the toolchain is the current JVM.

I think this needs to be made an `else if` with the block above.

hegyibalint replied:
Yeah, haven't seen this and realized myself :D
Had a deeper look at the diff, and implemented so the `isFork` branch always returns a spec.

hegyibalint commented on platforms/core-execution/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerDaemonLifecycleTest.groovy:
```diff
@@ -169,7 +169,7 @@ class WorkerDaemonLifecycleTest extends AbstractDaemonWorkerExecutorIntegrationS
         succeeds "compileJava", "runInWorker"
 
         then:
-        sinceSnapshot().count("Started Gradle worker daemon") == 2
```
I've checked, and the in-process spec is executed here; hence the one-less gradle worker

big-guy replied:
❌ This breaks what we're checking here.  This is supposed to be checking that when there are two workers (WorkerTask and java compilation), we stop the WorkerTask worker but not the java compilation one.

hegyibalint replied:
Okay, that means that there are two ways to do this:
 - A: By doing this change, as the java compiler becomes in-process
 - B: Forcing the java compiler to fork, so it will make a worker

I assume from your response that you would like to do option B.

big-guy replied:
correct

hegyibalint replied:
Now I noticed as you mentioned that the test explicitly asks for a worker, so something is probably wrong.

hegyibalint replied:
Okay, eventually nothing was "wrong" in the current HEAD.
The fix in the spec selection, which corrected the erroneous implementation in the `isForking` branch solved this automatically.

Dropped the commits doing work on this.

hegyibalint commented on platforms/jvm/language-java/src/main/java/org/gradle/api/internal/tasks/compile/AbstractJavaCompileSpecFactory.java:
```diff
@@ -16,53 +16,51 @@
     private final JavaInstallationMetadata toolchain;
 
-    public AbstractJavaCompileSpecFactory(CompileOptions compileOptions, @Nullable JavaInstallationMetadata toolchain) {
+    public AbstractJavaCompileSpecFactory(CompileOptions compileOptions, JavaInstallationMetadata toolchain) {
```
- `DefaultGroovyJavaJointCompileSpecFactory` didn't have a `@Nullable`
- `DefaultJavaCompileSpecFactory` had, and I've removed it

big-guy commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":null,"teamCityBuildId":null,"teamCityBuildHeadRef":"cc521959c4c47d8f2388309f3a14d1371e6d7514","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
I've triggered the following builds for you. Click [here](https://ge.gradle.org/scans?search.buildOutcome=failure&amp;search.names=gitCommitId&amp;search.tags=CI&amp;search.values=cc521959c4c47d8f2388309f3a14d1371e6d7514) to see all build failures.
- [PullRequestFeedback build](https://builds.gradle.org/build/84579051)

hegyibalint commented:
> @bot-gradle test this

I can see some breaking test, I will fix them.

bot-gradle commented:
<!-- {"replyTargetCommentId":2206541314,"teamCityBuildId":null,"teamCityBuildHeadRef":"cc521959c4c47d8f2388309f3a14d1371e6d7514","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->

Sorry I don't understand what you said.
Currently, the following commands are supported:
#### `@bot-gradle test <BuildTrigger1> <BuildTrigger2> ... <BuildTriggerN> [without PTS]`

Add `without PTS` to run the build with full test coverage.

- A trigger is a special build for this PR on TeamCity, common triggers are:
   - `SanityCheck`/`CompileAll`/`QuickFeedbackLinux`/`QuickFeedback`/`PullRequestFeedback`/`ReadyForNightly`/`ReadyForRelease`
   - Shortcuts: `SC`/`CA`/`QFL`/`QF`/`PRF`/`RFN`/`RFR`
   - Specific builds:
     - `BD`: `BuildDistributions`/`BuildDocs` so that you can preview the generated docs/distribution.
     - `PT`: `PerformanceTest`, all performance tests for Ready For Nightly stage.
     - `APT`: `AllPerformanceTest`, all performance tests, including slow performance tests.
     - `AST`: `AllSmokeTestsPullRequestFeedback`
     - `AFT`: `AllFunctionalTestsPullRequestFeedback`
     - `ASB`: `AllSpecificBuildsPullRequestFeedback`
     - `ACC`: `AllConfigCacheTestsPullRequestFeedback`
     - `ACT`: `AllCrossVersionTestsReadyForNightly`
     - `AFTN`: `AllFunctionalTestsReadyForNightly`
     - `ACTR`: `AllCrossVersionTestsReadyForRelease`
     - `AFTR`: `AllFunctionalTestsReadyForRelease`

#### `@bot-gradle squash`
- Squash the current pull request into a single commit. The squash message will come from the pull request body, with:
  - All `Signed-off-by:` lines kept.
  - All authors being `Co-Authored-By:`
  
#### `@bot-gradle merge`
- Enqueue this PR into GitHub merge queue for merging.
   - GitHub will create a merge commit from your PR branch HEAD and the target branch
   - A [GitHub Merge Queue Check Pass](https://builds.gradle.org/buildConfiguration/Gradle_Master_GitHubMergeQueueCheckPass) build will be triggered on the merge commit
   - When the build passes, the target branch will be fast-forwarded to this merge commit (i.e. merge the PR)

#### `@bot-gradle squash and merge`
- Squash the current pull request into a single commit as described in `squash` command above, 
  then enqueue this PR into GitHub merge queue as described in `merge` command above.

#### `@bot-gradle cherrypick to <branch>`

- Cherrypicks the current PR to another branch.
  A new PR will be created and a build will triggered automatically if there is no conflict.

#### `@bot-gradle cancel`

- cancel a running build in GitHub merge queue and remove the PR from the queue

#### `@bot-gradle clean`

- clear the conversation history

#### `@bot-gradle help`

- display this message

To run a command, simply submit a comment.  For detailed instructions see [here](https://bt-internal-docs.grdev.net/gbt/tools-services-and-checks/gradle-bot/).

hegyibalint commented on platforms/jvm/language-groovy/build.gradle.kts:
```diff
@@ -71,6 +71,12 @@ dependencies {
     integTestDistributionRuntimeOnly(project(":distributions-jvm"))
 }
 
+tasks.withType<Test>().configureEach {
+    if (!javaVersion.isJava9Compatible) {
+        classpath += javaLauncher.get().metadata.installationPath.files("lib/tools.jar")
+    }
+}
```
ℹ️ This existed already in the `language-java` project, and now caused a difference. The spec selection (through `JdkJavaCompiler.canBeUsed()` needs the `tools.jar` on the classpath to make the correct decision.

This is the default when we run the daemon itself, but this unit test didn't have that setup.

ljacomet reviewed (approved):
Discussed live, LGTM!

hegyibalint commented:
@bot-gradle merge
